video/transmux.go

```diff
@@ -95,21 +97,89 @@ func MuxTStoFMP4(fmp4ManifestOutputFile string, inputs ...string) error {
-	for _, k := range segmentsList.GetSortedSegments() {
-		segBytes, err := tsFile.Write(segmentsList.SegmentDataTable[k])
+	if !useStreamBasedConcat {
+		// Create a .ts file for a given rendition
+		tsFile, err := os.Create(tsFileName)
 		if err != nil {
-			return totalBytes, fmt.Errorf("error writing segment %d err: %w", k, err)
+			return totalBytes, fmt.Errorf("error creating file (%s) err: %w", tsFileName, err)
+		}
+		defer tsFile.Close()
+		// For a given rendition, write all segment indices in ascending order to the single .ts file
+		for _, k := range segmentsList.GetSortedSegments() {
+			segBytes, err := tsFile.Write(segmentsList.SegmentDataTable[k])
+			if err != nil {
+				return totalBytes, fmt.Errorf("error writing segment %d err: %w", k, err)
+			}
+			totalBytes = totalBytes + int64(segBytes)
+		}
+		return totalBytes, nil
+	} else {
+		// Strip the '.ts' from filename
+		fileBaseWithoutExt := tsFileName[:len(tsFileName)-len(".ts")]
+
+		// Create a text file containing filenames of the segments
+		segmentListTxtFileName := fileBaseWithoutExt + ".txt"
+		segmentListTxtFile, err := os.Create(segmentListTxtFileName)
+		if err != nil {
+			return totalBytes, fmt.Errorf("error creating  segment text file (%s) err: %w", segmentListTxtFileName, err)
+		}
+		defer segmentListTxtFile.Close()
+		defer os.Remove(segmentListTxtFileName)
+		w := bufio.NewWriter(segmentListTxtFile)
+
+		// Write each segment to disk and add segment filename to the text file
+		for segName, segData := range segmentsList.GetSortedSegments() {
+			// Open a new file to write each segment to disk
+			segmentFileName := fileBaseWithoutExt + "_" + strconv.Itoa(segName) + ".ts"
+			segmentFile, err := os.Create(segmentFileName)
+			if err != nil {
+				return totalBytes, fmt.Errorf("error creating individual segment file (%s) err: %w", segmentFileName, err)
+			}
+			defer segmentFile.Close()
+			defer os.Remove(segmentFileName)
+			// Write the segment data to disk
+			segBytes, err := segmentFile.Write(segmentsList.SegmentDataTable[segData])
+			if err != nil {
+				return totalBytes, fmt.Errorf("error writing segment %d err: %w", segName, err)
+			}
+			totalBytes = totalBytes + int64(segBytes)
+			// Add filename to the text file
+			line := fmt.Sprintf("file '%s'\n", segmentFileName)
+			if _, err = w.WriteString(line); err != nil {
+				return totalBytes, fmt.Errorf("error writing segment %d to text file err: %w", segName, err)
+			}
+			// Flush to make sure all buffered operations are applied
+			if err = w.Flush(); err != nil {
+				return totalBytes, fmt.Errorf("error flushing text file %s err: %w", segmentFileName, err)
+			}
 		}
-		totalBytes = totalBytes + int64(segBytes)
+		// Create a .ts file for a given rendition
+		tsFile, err := os.Create(tsFileName)
+		if err != nil {
+			return totalBytes, fmt.Errorf("error creating file (%s) err: %w", tsFileName, err)
+		}
+		defer tsFile.Close()
+		defer os.Remove(tsFileName)
```

Comment:
In the other disk-based pieces we did, we had logic on app startup to clear up any dangling files that got left behind from interrupted processes. Are we able to do that here?

Reply:
There's several deferred os.Removes in the top level fcn calls that ensures all dangling files are cleaned up if there's a panic/exit. But I didn't see any app startup logic to delete - can you point me to it?

Reply:
Yeah, we're not doing this for clip/thumbnails/transmux/etc. Discussed with @mjh1 on a way to do this better and I think we'll create a top level dir that we can remove from init() all at one. Filed a separate ticket to get this done (VID-474).